src/d/d_pad.cpp

```diff
@@ -0,0 +1,894 @@
+    }
+    return -1;
+}
+
+void ex_c::setInfo(s32 chan, const WPADInfo *pInfo) {
+    m_info[0][chan].dpd = pInfo->dpd;
+    m_info[0][chan].speaker = pInfo->speaker;
+    m_info[0][chan].attach = pInfo->attach;
+    m_info[0][chan].lowBat = pInfo->lowBat;
+    m_info[0][chan].nearempty = pInfo->nearempty;
+    m_info[0][chan].battery = pInfo->battery;
+    m_info[0][chan].led = pInfo->led;
+    m_info[0][chan].protocol = pInfo->protocol;
+    m_info[0][chan].firmware = pInfo->firmware;
+    m_connected[chan] = true;
+}
+
+f32 ex_c::fn_80058F50() {
+    return m_current_ex->mMotion.fn_80058A00();
+}
+
+bool ex_c::fn_80058F60() {}
+
+f32 ex_c::fn_80058FE0() {}
+
+void ex_c::fn_80058FF0() {}
+
+void ex_c::fn_80059000() {}
+
+void ex_c::fn_80059010() {}
+
+void ex_c::fn_800590A0() {}
+
+bool ex_c::fn_800590B0() {}
+
+bool ex_c::fn_800590E0() {}
+
+void ex_c::fn_800590F0() {}
+
+bool ex_c::fn_80059100() {
+    fn_80059110(mPad::g_currentCoreId);
+}
+
+bool ex_c::fn_80059110(s32 chan) {
+    return nw4r::math::VEC3LenSq(m_ex[chan].mMPLSVelocity) >= 0.01f || m_ex[chan].mFSStickDistance >= 0.0001f ||
+           mPad::g_core[chan]->getCoreStatus()->getHold() || mPad::g_core[chan]->getCoreStatus()->getTrigger() ||
+           m_ex[chan].mMotion.field_0x10C8 != 0;
+}
+
+void ex_c::fn_80059210() {
+    m_current_ex->field_0x22CE = 1;
+}
+
+void ex_c::fn_80059220() {
+    m_current_ex->field_0x22CF = 1;
+}
+
+void ex_c::fn_80059230() {
+    m_current_ex->field_0x22D0 = 1;
+}
+
+void ex_c::fn_80059240() {
+    m_current_ex->field_0x22D0 = 0;
+}
+
+void ex_c::on_0x54(s32 idx) {
+    m_ex[idx].field_0x54 = true;
+}
+void ex_c::on_0x55(s32 idx) {
+    m_ex[idx].field_0x55 = true;
+}
+void ex_c::on_0x56(s32 idx) {
+    m_ex[idx].field_0x56 = true;
+}
+void ex_c::on_0x57(s32 idx) {
+    m_ex[idx].field_0x57 = true;
+}
+
+void ex_c::fn_800592D0(s32 chan) {
+    m_ex[chan].field_0x54 = false;
+    m_ex[chan].field_0x55 = false;
+    m_ex[chan].field_0x56 = false;
+    m_ex[chan].field_0x57 = false;
+}
+
+void ex_c::fn_80059300(s32 chan) {
+    field_0x58 = field_0x54;
+    field_0x59 = field_0x55;
+    field_0x5A = field_0x56;
+    field_0x5B = field_0x57;
+    fn_800592D0(chan);
+}
+
+bool ex_c::fn_80059330(s32 chan) {
+    return m_ex[chan].field_0x58;
+}
+
+bool ex_c::fn_80059350(s32 chan) {
+    return m_ex[chan].field_0x59;
+}
+
+bool ex_c::fn_80059370(s32 chan) {
+    return m_ex[chan].field_0x5A;
+}
+
+bool ex_c::fn_80059390(s32 chan) {
+    return m_ex[chan].field_0x5B;
+}
+
+void ex_c::getUnifiedWpadStatus(s32 chan) {
+    KPADGetUnifiedWpadStatus(chan, &mStatus, 1);
+}
+
+void ex_c::fn_800593D0() {
+    field_0x22D1 = 0;
+    if (mFSStickDistance < 0.8f) {
+        field_0x22D4 = 0;
+        return;
+    }
+    u32 prev22D4 = field_0x22D4;
+    s32 ang = mFSStickAngle;
+    ang = ((ang + 0x11000) / 0x2000) % 8;
+    field_0x22D4 = 1 << ang;
+    if (field_0x22D4 != 0 && field_0x22D4 != prev22D4) {
+        field_0x22D1 = 1;
+    }
+}
```

Comment:
This turns the Nunchuk stick angle and distance into a "button mask" indicating which of the 8 directions it's pushed in, with `field_0x22D1` indicating whether the direction changed this frame.